Assets/MRTK/Services/InputSystem/FocusProvider.cs

```diff
@@ -1226,40 +1226,50 @@ private static void QueryScene(IMixedRealityPointer pointer, IMixedRealityRaycas
                                 Array.Resize<Collider>(ref colliders, maxQuerySceneResults);
                             }
 
-                            int numColliders = UnityEngine.Physics.OverlapSphereNonAlloc(pointer.Rays[i].Origin, pointer.SphereCastRadius, colliders, ~UnityEngine.Physics.IgnoreRaycastLayer);
-
-                            if (numColliders > 0)
+                            Vector3 testPoint = pointer.Rays[i].Origin;
+                            Vector3 objectHitPoint = testPoint;
+                            GameObject closest = null;
+                            float closestDistance = Mathf.Infinity;
+                            int totalNumColliders = 0;
+
+                            // Go through each layerMask and ensure perform the appropriate OverlapSphereCalculation
+                            // Since this is usually done when a pointer passes a IsInteractionEnabled, maybe we can cache the selected colliders inside the pointer?
+                            foreach (LayerMask layerMask in prioritizedLayerMasks)
                             {
-                                if (numColliders >= maxQuerySceneResults)
-                                {
-                                    Debug.LogWarning($"Maximum number of {numColliders} colliders found in FocusProvider overlap query. Consider increasing the focus query buffer size in the input profile.");
-                                }
-
-                                Vector3 testPoint = pointer.Rays[i].Origin;
-                                GameObject closest = null;
-                                float closestDistance = Mathf.Infinity;
-                                Vector3 objectHitPoint = testPoint;
-
-                                for (int colliderIndex = 0; colliderIndex < numColliders; colliderIndex++)
+                                int numColliders = UnityEngine.Physics.OverlapSphereNonAlloc(pointer.Rays[i].Origin, pointer.SphereCastRadius, colliders, layerMask);
+                                totalNumColliders += numColliders;
```

Comment:
For totalNumColliders, can you help explain a little more about why we're tracking the total number of things returned from OverlapSphereNonAlloc? I'm assuming that OverlapSphereNonAlloc uses the entire buffer (of size maxQuerySceneResults) instead of starting off at the previous totalNumColliders (i.e. I'm assuming it isn't just looking from the last set value in the array and honoring the fact that there are existing elements).

This type of function looks to be a general buffer passing one, which tend to not (at least, they really shouldn't) have memory of how much of the buffer is already used (usually when you pass a buffer you can assume that you can use the entire buffer).

Reply:
the totalNumColliders is mainly used to allow us to track and potentially issue the warning on lines 1243 through 1246. It's to ensure we have the same behavior as before when we only did 1 spherecast on  the one hardcoded layermask, but generalized to potentially multiple sphere casts on multiple different layermasks

Reply:
given that the numColliders queried per raycast is what we care to check, the counter for tracking the total number of colliders is a bit misleading. I'll clean this up